noise/README.md

```diff
@@ -0,0 +1,400 @@
+handshake that Bob was unable to decrypt, Bob will initiate the `XXfallback`
+handshake to Alice. This is reflected in the arrow direction above; fallback
+handshake patterns are notated in the [so-called][npf-alice-and-bob]
+"Bob-initiated form," with arrows reversed from the canonical (Alice-initiated)
+form.
+
+The handshake pattern is the same as in `XX`, however, Alice's ephemeral public
+key is obtained from her initial `IK` message, moving it to the pre-message
+section of the handshake pattern. Essentially, the failed `IK` message serves
+the same role as the first handshake message in the standard `XX` pattern.
+
+Both messages (not counting the initial failed `IK` message) include a
+[noise-libp2p handshake payload](#the-noise-libp2p-handshake-payload), which contains
+a signature authenticating the sender's static Noise key as described in the
+[Static Key Authentication section](#static-key-authentication).
+
+## Handshake Negotiation
+
+Peers agree upon which Noise handshake to use by leveraging libp2p's existing
+protocol negotiation system. At the time of writing, protocol negotiation uses
+multistream-select 1.0, as described in the [connection establishment
+spec][conn-spec], although [work is in progress][multiselect-2-pr] on defining
+a new version.
+
+### Protocol IDs
+
+A concrete Noise protocol is identified by the combination of handshake pattern,
+DH function, cipher, and hash function used. These elements combine to form a
+[Noise protocol name][npf-protocol-names]. 
+
+For example, the IK handshake using [the 25519 DH functions][npf-dh-25519],
+[ChaChaPoly cipher][npf-cipher-chachapoly] and [SHA-256 hash
+function][npf-hash-sha256] would have the Noise protocol name
+`Noise_IK_25519_ChaChaPoly_SHA256`.
+
+noise-libp2p protocol IDs have a similar format to Noise protocol names, adapted
+to libp2p naming conventions. A valid noise-libp2p protocol ID consists of the
+following elements, separated by `/` characters:
+
+- the string `/noise`
+- the name of the handshake pattern, lowercased
+- the name of the DH function, lowercased
+- the name of the cipher, lowercased
+- the name of the hash function, lowercased
+- the version of the noise-libp2p spec that defines the protocol
+
+If any of the names used to construct the protocol ID contains a `/` character
+(for example, `SHA3/256`), the `/` will be replaced with a `-` to avoid
+ambiguity with the separator character. In practice, none of the currently
+supported functions contain a `/` in their names, so this is likely to remain a
+non-issue.
+
+Using the [supported handshake patterns](#supported-handshake-patterns) and
+[supported DH function, cipher and
+hash](#supported-dh-functions-ciphers-and-hash-functions), the following are
+valid noise-libp2p protocol IDs:
+
+- `/noise/xx/25519/chachapoly/sha256/1.0.0`
+- `/noise/xxfallback/25519/chachapoly/sha256/1.0.0`
+- `/noise/ik/25519/chachapoly/sha256/1.0.0`
```

Comment:
Can someone please give some details of why the use of path-based protocol identifier is used? To me, the path here is not hierarchical, and also doesn't match the protocol identifier specified in the Noise protocol:
https://noiseprotocol.org/noise.html#protocol-names-and-modifiers
Why not use the Noise-specified protocol identifier after the "noise/" part of the path? The version can be used as the postfix.

Reply:
@shahankhatch I'm fine with using Noise names directly instead of wrangling them into paths like this. Are you thinking e.g. `/noise/IK_25519_ChaChaPoly_SHA256/1.0.0`?

Reply:
The Protocol Names and Modifiers section of the Noise protocol specification (linked above) says that the protocol name is something prefixed with "Noise_", like "Noise_IK_25519_ChaChaPoly_SHA256".
I think it's fine to have a path like `/noise/Noise_IX_25519_ChaChaPoly_SHA256/1.0.0` even if `Noise_` seems repetitive since it meets libp2p and Noise specs.

Reply:
In the #eth2 protocol we've decided not to go with semver for protocols due to the difficulties of interpreting what patch/minor upgrades mean and how clients should actually behave when they receive a not fully matching version - except maybe that if a client supports two versions, it might prioritise the newer protocol. use a single digit here also to avoid this confusion?

Reply:
@arnetheduck we haven't settled on the change yet in the libp2p community. The point has come up in several discussions, but it was lacking a dedicated one (I think). Until now. https://github.com/libp2p/specs/issues/203.